source/game/system/KPadController.cc

```diff
@@ -236,6 +301,87 @@ u8 KPadGhostTrickButtonsStream::readVal() const {
     return currentSequence >> 0x8 & ~0x80;
 }
 
+/* ================================ *
+ *     HOST CONTROLLER
+ * ================================ */
+
+KPadHostController::KPadHostController() = default;
+
+KPadHostController::~KPadHostController() = default;
+
+ControlSource KPadHostController::controlSource() {
+    return ControlSource::Host;
+}
+
+void KPadHostController::reset(bool driftIsAuto) {
+    m_driftIsAuto = driftIsAuto;
+    m_raceInputState.reset();
+    m_connected = true;
+}
+
+/// @brief Sets the inputs of the controller.
+/// @param state The specified inputs packaged in the state. Only buttons, stick, and trick matter.
+/// @return Input state validity.
+bool KPadHostController::setInputs(const RaceInputState &state) {
+    return setInputs(state.buttons, state.stick, state.trick);
+}
+
+/// @brief Sets the inputs of the controller.
+/// @param buttons The button inputs.
+/// @param stick The stick inputs, as a 2D vector.
+/// @param trick The trick input.
+/// @return Input state validity.
+bool KPadHostController::setInputs(u16 buttons, const EGG::Vector2f &stick, Trick trick) {
+    m_raceInputState.buttons = buttons;
+    m_raceInputState.stick = stick;
+    m_raceInputState.trick = trick;
+
+    return m_raceInputState.isValid();
+}
+
+/// @brief Sets the inputs of the controller.
+/// @param buttons The button inputs.
+/// @param stickX The stick input on the X axis.
+/// @param stickY The stick input on the Y axis.
+/// @param trick The trick input.
+/// @return Input state validity.
+bool KPadHostController::setInputs(u16 buttons, f32 stickX, f32 stickY, Trick trick) {
+    m_raceInputState.buttons = buttons;
```

Comment:
We need to be very careful here. We have these controller input validity checks, yet we are setting them in the state, regardless of their validity. AFAIK nothing here is undefined behavior in that scenario, but I do think it's interesting that we put in the effort to check (I guess at this point it's just for logging / up to the developer interfacing w/ Kinoko) yet we just continue on like usual.

Reply:
This is fine.